kernel/src/engine/arrow_expression.rs

```diff
@@ -303,6 +324,69 @@ fn evaluate_expression(
                 UnaryOperator::IsNull => Arc::new(is_null(&arr)?),
             })
         }
+        (
+            BinaryOperation {
+                op: In,
+                left,
+                right,
+            },
+            _,
+        ) => {
+            let left_arr = evaluate_expression(left.as_ref(), batch, None)?;
```

Comment:
Can't we optimize this quite easily?  

If `right` is a `Literal::(ArrayData)`, we can loop `left_arr` ourselves and probe `right`, possibly switching between array vs. hashtable based on the length of `left_arr`.

If `right` is a `Literal::(something_else)` I think that's an error and we can just return.

If `right` is an actual expression, we don't need to do the crazy exploding above because it'll just evaluate to _something_ and then we can use the `list_in` case.

Possibly I'm missing something, but that seems preferable to the "building of a repeated list column" above.

Reply:
You were thinking something similar to this?

```rust
match (left.as_ref(), right.as_ref()) {
  (_, Column(_)) => { // Use Arrow in_list },
  (Literal(lit), Literal(Scalar::Array(ad))) => { // Just do the comparison ourselves. },
  (_, _) => { // Some invalid comparison }
}
```

Reply:
My understanding is that we can encounter one of four cases:
* Literal value vs. array of literal values -- not really meaningful for data skipping since no column stats are implicated (and anyway the engine's query optimizer's constant expression eval should optimize it  away).
* Column vs. array of literal values -- the "standard" case, where I believe @nicklan was suggesting to manually search the array for each column value, rather than replicating the array N times.
* Column vs. uncorrelated subquery -- The engine would need to evaluate subqueries before triggering data skipping on the resulting array of literals. Might require some thought on how to integrate that with kernel's data skipping.
* Column vs. correlated subquery -- If the engine is unable to decorrelate the subquery into a join, it would need to evaluate subqueries and produce an arrow-style array of arrays. Probably not possible to integrate with data skipping, tho (because min and max column values are meaningless in this context).

Reply:
> You were thinking something similar to this?

Yep!

Regarding Ryan's comment. I was indeed discussing the second bullet point case.  

For uncorrelated subqueries, I think the engine could/would execute them independently and then pass that as just an array of literal values in an `IN` predicate.

Correlated subqueries are obviously trickier, but as Ryan notes we probably can't use data skipping there anyway.

Reply:
Maybe I misunderstood something? This code:
```rust
(Literal(lit), Literal(Scalar::Array(ad))) => { // Just do the comparison ourselves. },
```
would correspond to my first bullet point case `(<literal scalar> IN (<... literal values ...>)`, which could be evaluated at query compile time (no need to read anything off disk, not even stats).

The second bullet point, which I suggested we might want to optimize, would be 
```rust
(Column(...), Literal(Scalar::Array(ad))) => { // Just do the comparison ourselves. },
```
To handle this case, we'd need to grab the min/max stats for the column, and check whether they bracket any of the values in the array of scalars that represents the in-list.

Reply:
I think you understand this correctly, and I'm having a rough time coming up with a way to do column, literal that doesn't blow the code complexity up a massive amount. It's just that without some re-plumbing there doesn't look like a good way to feed column statistics into expression evaluation. Would it be acceptable to have an inefficient impl to start with and we can discuss a larger refactoring we'd might have to do to  support better handling of this?

Reply:
@scovich @nicklan I created #291 to discuss this further

Reply:
A couple reactions:

First, data skipping support for in-list is definitely an extra piece of work outside the scope of this PR, so we can defer the necessary stats plumbing to an issue and future work.

Second: The sub-optimal evaluation is (a) in our default engine's expression eval and (b) apparently The Arrow Way ("why that way?" being an obvious but out of scope question). So I agree we can defer being more clever, or possibly even choose not to be clever at all as long as we meet the "don't pessimize" bar. 

The bigger priority IMO is ensuring that kernel's general handling of IN-list allows engines to do a better job if they are willing to implement the expression eval themselves. That part gets trickier because engines have a lot of flexibility in how they rewrite and later eval such expressions. For example, I would hope most engines are smart enough to optimize `x in (10, 11)` as `x == 10 OR x == 11`, and some engines (IBM DB2 and sqlite3 among them) choose to convert oversized in-lists into a temp relation with an equijoin clause. In both cases, kernel would never even see an in-list. And for engines that do handle (medium-sized) IN-list directly, I would expect hash tables to figure prominently for their efficient probing. So we'll need to do some thinking about what kinds of in-list eval we expect to see, and try not to accidentally block engines from using their preferred approach.

Reply:
However, see my other comment below -- maybe we don't want kernel to expose an array literal type in the first place?